### PR TITLE
implemented delete and put schema api

### PR DESCRIPTION
**Pull request by ARADDCC002**

### CODE REVIEW

File: frontend/src/schemas/SchemaList.tsx [hunk 1]
```diff
@@ -1,35 +1,75 @@
...
+  const { schemas, isSchemasLoading, isSchemasError, mutateSchemas } = useGetSchemas(schemaKind)
+
+  const [errorMessage, setErrorMessage] = useState('')
+  const [open, setOpen] = useState(false)
+  const [schemaToBeDeleted, setSchemaToBeDeleted] = useState('')
+
+  const schemaList = useMemo(() => {
+    const handleSetSchemaActive = async (schema: SchemaInterface) => {
+      setErrorMessage('')
+      const updatedSchema = schema
```
Comment (a3957273) [inline frontend/src/schemas/SchemaList.tsx:25]: This is a reference copy, it is a useless statement.

**Review by a3957273 — APPROVED** [review]

File: backend/src/routes/v2/schema/deleteSchema.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+      required_error: 'Must specify schema id as URL parameter',
+    }),
+  }),
+})
+
+registerPath({
+  method: 'delete',
+  path: '/api/v2/schema/{schemaId}',
+  tags: ['schema'],
+  description: 'Update an existing schema.',
```
Comment (JR40159) [inline backend/src/routes/v2/schema/deleteSchema.ts:23]: Copy and paste error?

File: backend/src/routes/v2/schema/deleteSchema.ts [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+})
+
+interface DeletedSchemaResponse {
+  deleted: boolean
+}
+
+export const deleteSchema = [
+  bodyParser.json(),
+  async (req: Request, res: Response<DeletedSchemaResponse>) => {
+    req.audit = AuditInfo.CreateSchema
```
Comment (JR40159) [inline backend/src/routes/v2/schema/deleteSchema.ts:46]: Incorrect audit event

File: backend/src/routes/v2/schema/putSchema.ts
Comment (JR40159) [inline backend/src/routes/v2/schema/putSchema.ts:1]: Allow a user to update any part of a schema seems unnecessarily risky. Updating a schema could cause problems with models and access requests that have previously been created using that schema. If a user wants to update a schema, they could (and should?) create a new version of the schema and mark the previous one as inactive.

File: backend/src/routes/v2/schema/putSchema.ts [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+})
+
+interface PutSchemaResponse {
+  schema: SchemaInterface
+}
+
+export const putSchema = [
+  bodyParser.json(),
+  async (req: Request, res: Response<PutSchemaResponse>) => {
+    req.audit = AuditInfo.CreateSchema
```
Comment (JR40159) [inline backend/src/routes/v2/schema/putSchema.ts:61]: Incorrect audit event

File: backend/src/routes/v2/schema/putSchema.ts [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    const { body } = parse(req, putSchemaSchema)
+
+    const schema = await createSchema(req.user, body, true)
```
Comment (JR40159) [inline backend/src/routes/v2/schema/putSchema.ts:64]: Create?

File: frontend/src/schemas/SchemaList.tsx
Comment (JR40159) [inline frontend/src/schemas/SchemaList.tsx:1]: Do we have/need a warning in the UI about the possible consequences of deleting a schema?

**Review by JR40159 — CHANGES_REQUESTED** [review]

File: backend/src/services/v2/schema.ts [hunk 6]
```diff
@@ -25,6 +25,28 @@ export async function findSchemaById(schemaId: string): Promise<SchemaInterface>
...
+
+  const auth = await authorisation.schema(user, schema, SchemaAction.Delete)
+  if (!auth.success) {
+    throw Forbidden(auth.info, {
+      userDn: user.dn,
+      schemaId: schema.id,
+    })
+  }
+
+  await schema.remove()
```
Comment (JR40159) [inline backend/src/services/v2/schema.ts:45]: It looks like `.remove()` has been deprecated.

File: frontend/src/schemas/SchemaList.tsx [hunk 7]
```diff
@@ -1,35 +1,75 @@
...
-      )),
-    [schemas],
-  )
+  const { schemas, isSchemasLoading, isSchemasError, mutateSchemas } = useGetSchemas(schemaKind)
+
+  const [errorMessage, setErrorMessage] = useState('')
+  const [open, setOpen] = useState(false)
+  const [schemaToBeDeleted, setSchemaToBeDeleted] = useState('')
+
+  const schemaList = useMemo(() => {
```
Comment (ARADDCC012) [inline frontend/src/schemas/SchemaList.tsx:22]: memo hooks shouldn't contain side effects. I'd recommend moving most of this logic into a new component and using the memo to only handle managing a list of those components.

File: frontend/src/schemas/SchemaList.tsx [hunk 8]
```diff
@@ -1,35 +1,75 @@
...
+      updatedSchema.active = !schema.active
+      const res = await putSchema(updatedSchema)
+      if (!res.ok) {
+        setErrorMessage(await getErrorMessage(res))
+      } else {
+        mutateSchemas()
+      }
+    }
+
+    const handleDeleteSchemaButtonOnClick = (schemaId) => {
```
Comment (ARADDCC012) [inline frontend/src/schemas/SchemaList.tsx:35]: `schemaId` should have a type

File: frontend/src/schemas/SchemaList.tsx [hunk 9]
```diff
@@ -1,35 +1,75 @@
...
+      setOpen(true)
+      setSchemaToBeDeleted(schemaId)
+    }
+
+    const handleDeleteConfirm = async (schema: string) => {
```
Comment (ARADDCC012) [inline frontend/src/schemas/SchemaList.tsx:40]: `schema` suggests this is the schema object. Should this be `schemaId`?

File: frontend/src/schemas/SchemaList.tsx
Comment (ARADDCC012) [reply frontend/src/schemas/SchemaList.tsx:1]: `ConfirmationDialog` could be updated to have an optional `message` to provide this sort of information if needed.

**Review by ARADDCC012 — CHANGES_REQUESTED** [review]

File: backend/src/routes/v2/schema/putSchema.ts (hunk 5 above)
Comment (ARADDCC002) [reply backend/src/routes/v2/schema/putSchema.ts:64]: I believe we discussed this and we were happy to re-use the `createSchema` function as it has an existing override property. We can update this if you're not happy, though.

File: frontend/src/schemas/SchemaList.tsx [hunk 10]
```diff
@@ -22,24 +22,23 @@ export default function SchemaList({ schemaKind }: SchemaDisplayProps) {
   const schemaList = useMemo(() => {
     const handleSetSchemaActive = async (schema: SchemaInterface) => {
       setErrorMessage('')
-      const updatedSchema = schema
-      updatedSchema.active = !schema.active
-      const res = await putSchema(updatedSchema)
+      schema.active = !schema.active
```
Comment (ARADDCC012) [inline frontend/src/schemas/SchemaList.tsx:25]: We shouldn't be mutating parameters. It's safer to do:
```
const updatedSchema = { ...schema, active: !schema.active }
```

**Review by ARADDCC012 — APPROVED** [review]

**Review by JR40159 — APPROVED** [review]

**Pull request merged**